lib/storage-client.js

```diff
@@ -0,0 +1,345 @@
+    const childNode = parent.childNodes[childNodeIdx];
+    if (childName && !text && childName === childNode.localName) {
+      return childNode;
+    }
+    if (text) {
+      const childText = extractNodeText(childNode);
+      if (!childText) {
+        continue;
+      }
+      if (childName && childName === childNode.localName && text === childText) {
+        return childNode;
+      }
+      if (!childName && text === childText) {
+        return childNode;
+      }
+    }
+  }
+  return null;
+}
+
+function extractNodeText (node) {
+  if (!node || !node.firstChild || !util.hasValue(node.firstChild.nodeValue)) {
+    return null;
+  }
+  return node.firstChild.nodeValue;
+}
+
+
+class ChromedriverStorageClient {
+  constructor (args = {}) {
+    const {
+      chromedriverDir = getChromedriverDir(),
+      timeout = TIMEOUT,
+    } = args;
+    this.chromedriverDir = chromedriverDir;
+    this.timeout = timeout;
+    this.mapping = {};
+  }
+
+  parseNotes (content) {
+    const versionMatch = /^\s*[-]+ChromeDriver[\D]+([\d.]+)/im.exec(content);
+    if (!versionMatch) {
+      return {};
+    }
+    const minBrowserVersionMatch = /^\s*Supports Chrome[\D]+(\d+)/im.exec(content);
+    if (!minBrowserVersionMatch) {
+      return {
+        version: versionMatch[1],
+      };
+    }
+    return {
+      version: versionMatch[1],
+      minBrowserVersion: minBrowserVersionMatch[1],
+    };
+  }
+
+  async parseStorageXml (doc, parseNotes = true) {
+    const driverNodes = xpath.select(`//*[local-name(.)='Contents']`, doc);
+    log.debug(`Parsed ${driverNodes.length} entries from storage XML`);
+    if (_.isEmpty(driverNodes)) {
+      return;
+    }
+
+    const threads = [];
+    for (const driverNode of driverNodes) {
+      const key = extractNodeText(findChildNode(driverNode, 'Key'));
+      if (!_.includes(key, '/chromedriver_')) {
+        continue;
+      }
+
+      log.debug(`Processing chromedriver entry '${key}'`);
+      const etag = extractNodeText(findChildNode(driverNode, 'ETag'));
+      if (!etag) {
+        log.debug(`The entry '${key}' does not contain the checksum. Skipping it`);
+        continue;
+      }
+
+      const cdInfo = {
+        url: `${STORAGE_URL}/${key}`,
+        etag: _.trim(etag, '"'),
+        version: _.first(key.split('/')),
+      };
+      this.mapping[key] = cdInfo;
+      log.debug(`Added a new entry to the storage mapping: ${JSON.stringify(cdInfo)}`);
+
+      const notesPath = `${cdInfo.version}/notes.txt`;
+      const isNotesPresent = !!driverNodes
+        .reduce((acc, node) => acc || findChildNode(node, 'Key', notesPath), false);
+      if (!isNotesPresent) {
+        cdInfo.minBrowserVersion = null;
+        if (parseNotes) {
+          log.info(`The entry '${key}' does not contain any notes. Skipping it`);
+        }
+        continue;
+      } else if (!parseNotes) {
+        continue;
+      }
+
+      threads.push((async () => {
```

Comment:
Could we clean this part a bit and maybe turn it into methods? Threads and promises all together looks a but confusing

Reply:
refactored